design/014-secretstore-generator-v2.md

```diff
@@ -0,0 +1,198 @@
+
+* SecretStores support ClusterSecretStores with Referrent Authentication and cross namespace generation; Generators dont;
+* Generators support GeneratorState to track down eventual cache information on the CRD; SecretStores dont;
+* SecretStores Are periodically refreshed; Generators arent;
+* Badly configured SecretStores prevent jobs to happen via a gating mechanism; Generators dont;
+
+## Proposal
+
+The proposal is a new CRD for both SecretStores and Generators that unifies all of their feature sets, bringing the best of the worlds. Since this is a significant change, the idea is to also leverage a new controller and client manager for this CRD, in order to make it easier to maintain and scale.
+
+While changing the provider interface for both SecretStores and Generators itself is a non goal, it might be needed to accomodate this proposal.
+
+### SecretStore
+SecretStore manifests looks very much like our original implementation. Provider configuration resides on `spec.provider` field. 
+```yaml
+apiVersion: secretstore.external-secrets.io/v2alpha1
+kind: AWSSecretManager
+metadata:
+  name: my-aws-store
+  namespace: default
+spec:
+  refreshInterval: 1m
+  controller: dev
+  retrySettings:
+    maxRetries: 3
+    retryInterval: 10s
+  providerConfig:
+    type: InTree/OutOfTree
+    address: http+unix:///path/to/socket.sock
+  cachingPolicy: ProviderDefined/Secrets/Auth/All/None
+  cacheConfig: # if Caching Policy is Secrets/Auth/All 
+      ttl: 1m
+      maxEntries: 100
+  gatingPolicy: Enabled/Disabled
+  provider: ## same as current secretstore.Provider field
+      role: arn:aws:iam::123456789012:role/external-secrets
+      region: eu-central-1
+      auth:
+        secretRef:
+          accessKeyIDSecretRef:
+            name: awssm-secret
+            key: access-key
+            # note: namespace key in here is always blank. and removed from manifests
+          secretAccessKeySecretRef:
+            name: awssm-secret
+            key: secret-access-key    
+```
+New fields would be:
+* gatingPolicy - per SecretStore, explicit behavior for `enablefloodGate` flag (currently at controller level only)
+* cachingPolicy - as to document per SecretStore what's the caching strategy (cachingScope, ttl, maxEntries). As some secretStores do not have this control in place. `ProviderDefined` policy means whatever the provider defines as the best practice.
+* providerConfig (this name is terrible) - whether this is an InTree or OutOfTree provider. if OutOfTree, we don't expect any of the `Provider` Interface to be implemented, and instead just delegate to the outOfTree provider that a new job needs to be executed to fetch a given Secret. It is the provider responsibility to then Read the CRDs and any other references to it to reply with a valid `map[string][]byte`
```

Comment:
Think of it as delegating the `GetProviderSecretData` method for that specific provider instead of wanting a `Provider`  Interface  to exist for this CR.

Reply:
You didn't describe what address would be? Is that to a binary? A unix socket would mean that something is literally running inside the pod or has access to a shared volume.

Reply:
Ah! comment is here!

Yeah, it would be to a binary, mounted via a shared volume via CSI/daemonset. Im not that concerned with this one as it is really just to support safe out-of-tree plugins, not to move our own stuff to it. So it is a better extension point for it (one that we lack because of the current Provider Interface (that does not get the object, just gets the `remoteRef`).

Reply:
thats a possibility. Though, how would point out our own intree code?

I mean, that would mean we need to move it all outOfTree right?